From c52a967e48c3d91ff58c77ec7f587da216dc9978 Mon Sep 17 00:00:00 2001
From: Baha Mesleh <baha.mesleh@nokia.com>
Date: Thu, 18 Mar 2021 11:47:37 +0000
Subject: [PATCH 1344/1921] octeontx2-bphy-netdev: fix cleanup sequence in char
 device release

This patch fixes the kernel panic caused by receiving packets
during the cleanup of the resources used by the driver as a
result of calling otx2_bphy_cdev_release after the crash of
the odp application.

Change-Id: I8f217f66d185de89bf1e28ad82391e526ad9bbcf
Signed-off-by: Baha Mesleh <baha.mesleh@nokia.com>
Signed-off-by: Naveen Mamindlapalli <naveenm@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/49735
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <SunilKovvuri.Goutham@cavium.com>
(cherry picked from commit 4a9c2f027f97cd33eb06dd81cc1b030360b4df1a)
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/49830
Tested-by: Sunil Kovvuri Goutham <SunilKovvuri.Goutham@cavium.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 .../marvell/octeontx2/bphy/otx2_bphy_main.c   | 25 +++++++++++++------
 1 file changed, 18 insertions(+), 7 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/bphy/otx2_bphy_main.c b/drivers/net/ethernet/marvell/octeontx2/bphy/otx2_bphy_main.c
index 9e8f4d033a56..f485ed825bc5 100644
--- a/drivers/net/ethernet/marvell/octeontx2/bphy/otx2_bphy_main.c
+++ b/drivers/net/ethernet/marvell/octeontx2/bphy/otx2_bphy_main.c
@@ -182,12 +182,18 @@ static long otx2_bphy_cdev_ioctl(struct file *filp, unsigned int cmd,
 	}
 	case OTX2_RFOE_IOCTL_ODP_DEINIT:
 	{
-		otx2_bphy_rfoe_cleanup();
-		otx2_bphy_cpri_cleanup();
+		u32 status;
 
 		/* Disable GPINT Rx and Tx interrupts */
-		writeq(0xFFFFFFFF,
-		       bphy_reg_base + PSM_INT_GP_ENA_W1C(1));
+		writeq(0xFFFFFFFF, bphy_reg_base + PSM_INT_GP_ENA_W1C(1));
+
+		/* clear interrupt status */
+		status = readq(bphy_reg_base + PSM_INT_GP_SUM_W1C(1)) &
+				0xFFFFFFFF;
+		writeq(status, bphy_reg_base + PSM_INT_GP_SUM_W1C(1));
+
+		otx2_bphy_rfoe_cleanup();
+		otx2_bphy_cpri_cleanup();
 
 		cdev->odp_intf_cfg = 0;
 
@@ -498,18 +504,23 @@ static int otx2_bphy_cdev_open(struct inode *inode, struct file *filp)
 static int otx2_bphy_cdev_release(struct inode *inode, struct file *filp)
 {
 	struct otx2_bphy_cdev_priv *cdev = filp->private_data;
+	u32 status;
 
 	mutex_lock(&cdev->mutex_lock);
 
 	if (!cdev->odp_intf_cfg)
 		goto cdev_release_exit;
 
-	otx2_bphy_rfoe_cleanup();
-	otx2_bphy_cpri_cleanup();
-
 	/* Disable GPINT Rx and Tx interrupts */
 	writeq(0xFFFFFFFF, bphy_reg_base + PSM_INT_GP_ENA_W1C(1));
 
+	/* clear interrupt status */
+	status = readq(bphy_reg_base + PSM_INT_GP_SUM_W1C(1)) & 0xFFFFFFFF;
+	writeq(status, bphy_reg_base + PSM_INT_GP_SUM_W1C(1));
+
+	otx2_bphy_rfoe_cleanup();
+	otx2_bphy_cpri_cleanup();
+
 	cdev->odp_intf_cfg = 0;
 
 cdev_release_exit:
-- 
2.31.1

